### PR TITLE
Fix race condition: Protect access to Kafka decoder callbacks with a mutex

### PR DESCRIPTION
**Pull request by martyngigg**
**Description of work.**

Helgrind picked up a data race in the `registerCb` functions of `IKafkaStreamDecoder`. This has been causing sporadic failures of the [`KafkaHistoStreamDecoderTest`](http://builds.mantidproject.org/job/pull_requests-ubuntu/25866/console).

The callback functions are now protected by a mutex so that they can be reset while the stream capture is still in progress.

**To test:**

Code review to understand the origin and fix.

*There is no associated issue.*

*This does not require release notes* because it is an internal change.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/LiveData/inc/MantidLiveData/Kafka/IKafkaStreamDecoder.h [hunk 1]
```diff
@@ -27,7 +27,32 @@ functionality for accessing the data stream and processing data.
...
+
+    Callback(Callback::FnType callback) : m_mutex(), m_callback() {
+      setFunction(std::move(callback));
+    }
+
+    inline void operator()() {
+      std::lock_guard<std::mutex> lck(m_mutex);
+      m_callback();
+    }
+    void setFunction(Callback::FnType callback) {
```
Comment (quantumsteve) [inline Framework/LiveData/inc/MantidLiveData/Kafka/IKafkaStreamDecoder.h]: Using std::move on a member function is a code smell.  Can this be passed by const ref? See, for example, https://youtu.be/xnqTKD8uD64

File: Framework/LiveData/inc/MantidLiveData/Kafka/IKafkaStreamDecoder.h [hunk 2]
```diff
@@ -57,10 +82,12 @@ class DLLExport IKafkaStreamDecoder {
 
   ///@name Callbacks
   ///@{
-  virtual void registerIterationEndCb(CallbackFn cb) {
-    m_cbIterationEnd = std::move(cb);
+  virtual void registerIterationEndCb(Callback::FnType cb) {
```
Comment (quantumsteve) [inline Framework/LiveData/inc/MantidLiveData/Kafka/IKafkaStreamDecoder.h]: Using std::move on a member function is a code smell. Can this be passed by const ref?

File: Framework/LiveData/inc/MantidLiveData/Kafka/IKafkaStreamDecoder.h [hunk 3]
```diff
@@ -27,7 +27,32 @@ functionality for accessing the data stream and processing data.
...
+   * Defines a thread-safe callback. A mutex is held
+   * both during assignment of a new callback function
+   * and during the call.
+   */
+  class Callback {
+  public:
+    using FnType = std::function<void()>;
+
+    Callback(Callback::FnType callback) : m_mutex(), m_callback() {
+      setFunction(std::move(callback));
```
Comment (quantumsteve) [inline Framework/LiveData/inc/MantidLiveData/Kafka/IKafkaStreamDecoder.h]: I believe constructors are the exception.

https://clang.llvm.org/extra/clang-tidy/checks/modernize-pass-by-value.html

**Review by quantumsteve — COMMENTED** [review]
Saw a few code smells we may want to address here or in a separate PR.

**Review by peterfpeterson — DISMISSED** [review]
Adding a mutex is a reasonable solution.

I ran `ctest -R Kafka --repeat-until-fail 100` without issue. This took a bit as I didn't expect `LiveDataTest_KafkaEventStreamDecoderTest` to take 6-22s per run.

**Review by quantumsteve — APPROVED** [review]

**Pull request merged**